server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java

```diff
@@ -333,6 +346,33 @@ public void setHasPendingAsyncFetch() {
         this.hasPendingAsyncFetch = true;
     }
 
+    public long unaccountableSearchableSnapshotSize(RoutingNode node) {
+        try {
+            return unaccountableSearchableSnapshotSizes.computeIfAbsent(node, k -> {
+                long totalSize = 0;
+                // Count the STARTED searchable snapshot shards which are unaccounted in the cluster
+                for (ShardRouting shard : node.started()) {
+                    if (clusterInfo == null) {
+                        continue;
+                    }
+                    DiskUsage usage = clusterInfo.getNodeMostAvailableDiskUsages().get(node.nodeId());
+                    ClusterInfo.ReservedSpace reservedSpace = clusterInfo.getReservedSpace(
+                        node.nodeId(),
+                        usage != null ? usage.getPath() : ""
+                    );
+                    if (clusterState.metadata().getIndexSafe(shard.index()).isSearchableSnapshot()
+                        && reservedSpace.containsShardId(shard.shardId()) == false
+                        && clusterInfo.getShardSize(shard) == null) {
+                        totalSize += Math.max(shard.getExpectedShardSize(), 0L);
+                    }
+                }
+                return totalSize;
+            });
+        } catch (ExecutionException ignore) {
```

Comment:
Once we make `unaccountableSearchableSnapshotSizes` non-static, we can use a `Map` instead and I think this exception is then no longer needed. I'd prefer not to swallow exceptions.

Reply:
True that!